store/store.go

```diff
@@ -546,6 +559,110 @@ type requestOptions struct {
 	Data         []byte
 }
 
+type httpErrorResponse struct {
+	StatusCode int
+	URL        *url.URL
+	Method     string
+	Oops       string
+}
+
+func NewHttpErrorResponse(resp *http.Response) *httpErrorResponse {
+	return &httpErrorResponse{StatusCode: resp.StatusCode, URL: resp.Request.URL, Method: resp.Request.Method}
+}
+
+func (r httpErrorResponse) Error() string {
+	return fmt.Sprintf("HTTP request failed: '%v', method: %s, status: %d", r.URL, r.Method, r.StatusCode)
+}
+
+func (s *Store) doStoreRequest(client *http.Client, req *http.Request, user *auth.UserState, dataDecodeFunc func(*http.Response) error) error {
+	var err error
+	for refreshed := true; refreshed; {
+		refreshed = false
+
+		resp, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+
+		wwwAuth := resp.Header.Get("WWW-Authenticate")
+		if resp.StatusCode == http.StatusUnauthorized {
+			if user != nil && strings.Contains(wwwAuth, "needs_refresh=1") {
+				// refresh user
+				err := s.refreshUser(user)
+				if err != nil {
+					resp.Body.Close()
+					return err
+				}
+				refreshed = true
+			}
+			if strings.Contains(wwwAuth, "refresh_device_session=1") {
+				// refresh device session
+				if s.authContext == nil {
+					resp.Body.Close()
+					return fmt.Errorf("internal error: no authContext")
+				}
+				device, err := s.authContext.Device()
+				if err != nil {
+					resp.Body.Close()
+					return err
+				}
+
+				err = s.refreshDeviceSession(device)
+				if err != nil {
+					resp.Body.Close()
+					return err
+				}
+				refreshed = true
+			}
```

Comment:
Do we have a test for when it's unauthorised but none of these cases? Shouldn't that also return an error?